functest/extract_test.go

```diff
@@ -0,0 +1,198 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package functest
+
+import (
+	"io/ioutil"
+	"log"
+	"os"
+	"path/filepath"
+	"reflect"
+	"regexp"
+	"strings"
+	"testing"
+
+	cmdtest "github.com/appsody/appsody/cmd/cmdtest"
+)
+
+// Simple test for appsody extract command.
+// A future enhancement is to extract with buildah.
+// I could have very well taken one single stack
+// and test, but due to the fact that most pain points
+// come from diverse mounts that exist for different
+// stacks, it is nice to loop through each and make sure
+// nothing is broken.
+
+func TestExtract(t *testing.T) {
+
+	log.Println("stacksList is: ", stacksList)
+
+	// if stacksList is empty there is nothing to test so return
+	if stacksList == "" {
+		log.Println("stacksList is empty, exiting test...")
+		return
+	}
+
+	// replace incubator with appsodyhub to match current naming convention for repos
+	stacksList = strings.Replace(stacksList, "incubator", "appsodyhub", -1)
+
+	// split the appsodyStack env variable
+	stackRaw := strings.Split(stacksList, " ")
+
+	// loop through the stacks
+	for i := range stackRaw {
+
+		// create a temporary dir to create the project
+		projectDir, err := ioutil.TempDir("", "appsody-extract-test-"+strings.ReplaceAll(stackRaw[i], "/", "_"))
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		defer os.RemoveAll(projectDir)
+		log.Println("Created project dir: " + projectDir)
+
+		// create a temporary dir to extract the project, sibling to projectDir
+		parentDir := filepath.Dir(projectDir)
+		if err != nil {
+			t.Fatal(err)
+		}
+		extractDir := parentDir + "/appsody-extract-test-extract-" + strings.ReplaceAll(stackRaw[i], "/", "_")
+
+		defer os.RemoveAll(extractDir)
+		log.Println("Created extraction dir: " + extractDir)
+
+		// appsody init inside projectDir
+		log.Println("Now running appsody init...")
+		_, err = cmdtest.RunAppsodyCmdExec([]string{"init", stackRaw[i]}, projectDir)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		// appsody extract: running in projectDir, extracting into extractDir
+		var stdout string
+		stdout, err = cmdtest.RunAppsodyCmdExec([]string{"extract", "--target-dir", extractDir, "-v"}, projectDir)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		// Main extraction test logic:
+		// 1. Don't inspect the images, but just figure out the mount points
+		// from the debug log that is thrown out from extraction
+		// 2. For each mount that is a file, make sure the source and
+		// destination file sizes match.
+		// 3. For each mount that is a folder, make sure the source and
+		// destination folders have same content (file name match)
+		// 4. Skip mounts that were not extracted (e.g:- /mvn/...)
+		// 5. For all cases, make sure we have a Dockerfile extracted
+		// at the vortex of the extraction.
+		r, _ := regexp.Compile("APPSODY_MOUNTS Value: .*")
```

Comment:
What's the rationale behind using the debug output to find out what the mounts are? We like the test, but at the same time we think it's risky to depend on a debug message to figure out the mounts.

Reply:
@chilanti - I was concerned about this too! 

my rationale: if we want to obtain the environment of the stack, I would need to either do almost everything as `extract` do, or invoke APIs that take part in extraction logic. In ether way, any potential bug that affects `extract` can also infect into the test.

I changed it now, and took a new approach - basically run the `extract` in-process, temporarily switch the working folder the project location, get the environment through `GetEnvVar` - right now this is not exported, but a change is proposed to that effect through https://github.com/appsody/appsody/pull/322 , pls have a look!

Reply:
Thanks - will take a look - and I do see your point (it doesn't make sense to test `extract` with some of the same `extract` logic). But we'll take it.